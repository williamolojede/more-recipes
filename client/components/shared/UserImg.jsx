import React from 'react';

import { userImgPropTypes } from '../../config/proptypes';

const UserImg = ({ user, type }) => {
  const buffer = user.fullname.split(' ');
  let firstChar;
  let secondChar;

  if (buffer.length > 1) {
    firstChar = buffer[0].charAt(0).toUpperCase();
    secondChar = buffer[1].charAt(0).toUpperCase();
  } else {
    firstChar = buffer[0].charAt(0).toUpperCase();
  }

  return (
    <div className={`user-image ${type}`}>
      {
        !user.imgUrl ?
          <div className="user-image__no-img">
            <span>{firstChar}</span>
            <span>{secondChar}</span>
          </div>
          :
          <img src={user.imgUrl} alt={user.fullname} title={user.fullname} className="user-image__img" />
      }
    </div>
  );
};

UserImg.propTypes = userImgPropTypes;

export default UserImg;
