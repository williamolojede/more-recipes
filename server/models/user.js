import bcrypt from 'bcrypt';

module.exports = (sequelize, DataTypes) => {
  const User = sequelize.define('User', {
    username: {
      allowNull: true,
      unique: true,
      validate: {
        max: 15,
      },
      type: DataTypes.STRING,
    },
    email: {
      allowNull: false,
      unique: true,
      validate: {
        isEmail: true,
      },
      type: DataTypes.STRING,
    },
    fullname: {
      allowNull: false,
      validate: {
        max: 50,
      },
      type: DataTypes.STRING,
    },
    password: {
      allowNull: false,
      type: DataTypes.STRING,
      set(password) {
        const hash = bcrypt.hashSync(password, 10);
        this.setDataValue('password', hash);
      }
    },
  });

  User.associate = (models) => {
    User.hasMany(models.Recipe, {
      foreignKey: 'owner',
      as: 'recipes',
    });

    User.hasMany(models.Favorite, {
      foreignKey: 'userId',
      as: 'favorites',
    });

    User.hasMany(models.Review, {
      foreignKey: 'userId',
      as: 'reviews',
    });
  };

  User.authenticate = (email, password, callback) => {
    // console.log(email, password);
    User.findOne({
      where: { email },
    })
      .then((user) => {
        // check if it finds no user
        if (!user) {
          const err = new Error('User not found');
          err.status = 404;
          return callback(err);
        }
        const isMatch = bcrypt.compareSync(password, user.password);

        // check if password matches
        if (isMatch) return callback(null, user);
        return callback();
      })
      .catch(err => callback(err));
  };
  return User;
};
