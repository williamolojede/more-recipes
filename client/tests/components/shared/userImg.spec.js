import React from 'react';
import UserImg from '../../../components/shared/UserImg';
import user from '../../__mocks__/user.mock';
import defaultConfig from '../../../config/default';

function setup() {
  const props = {
    user,
    type: 'inReview'
  };
  const wrapper = shallow(<UserImg {...props} />);

  return {
    wrapper,
    props
  };
}

describe('Top Rated Recipes List', () => {
  it('should render with right amount of elements', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render properly if user fullname is just one word', () => {
    const { wrapper, props } = setup();
    const propsWithUserImgUrl = Object.assign({}, props);
    propsWithUserImgUrl.user.fullname = 'OneWord';
    wrapper.setProps({ propsWithUserImgUrl });
    expect(wrapper).toMatchSnapshot();
  });

  it('should render an img tag if user imgurl is present', () => {
    const { wrapper, props } = setup();
    const propsWithUserImgUrl = Object.assign({}, props);
    propsWithUserImgUrl.user.imgUrl = defaultConfig.userImgUrl;
    wrapper.setProps({ propsWithUserImgUrl });
    expect(wrapper.find('img').length).toBe(1);
    expect(wrapper).toMatchSnapshot();
  });
});
