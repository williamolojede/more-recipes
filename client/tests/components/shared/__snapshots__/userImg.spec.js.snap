// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Top Rated Recipes List should render an img tag if user imgurl is present 1`] = `
<div
  className="user-image inReview"
>
  <img
    alt="OneWord"
    className="user-image__img"
    src="https://www.worldcrunch.com/assets/img/avatars/thumbnails/default-user-img-profile.jpg"
    title="OneWord"
  />
</div>
`;

exports[`Top Rated Recipes List should render properly if user fullname is just one word 1`] = `
<div
  className="user-image inReview"
>
  <div
    className="user-image__no-img"
  >
    <span>
      O
    </span>
    <span />
  </div>
</div>
`;

exports[`Top Rated Recipes List should render with right amount of elements 1`] = `
<div
  className="user-image inReview"
>
  <div
    className="user-image__no-img"
  >
    <span>
      J
    </span>
    <span>
      D
    </span>
  </div>
</div>
`;
